### PR TITLE
Fix autobahn No. 2.5

### PR DESCRIPTION
**Pull request by riemass**
Ping (and control frames in general) can't cant have a payload larger than 125 bytes. Also, shutdown exit code in this case should be 1002 (protocol error).

### CODE REVIEW

File: libcaf_net/src/detail/rfc6455.cpp [hunk 1]
```diff
@@ -84,6 +84,10 @@ ptrdiff_t rfc6455::decode_header(const_byte_span data, header& hdr) {
   bool masked = (byte2 & 0x80) != 0;
   auto len_field = byte2 & 0x7F;
   size_t header_length;
+  // control frames can only have payload up to 125 bytes
+  if (rfc6455::is_control_frame(hdr.opcode) && len_field > 125) {
+    return -1;
+  }
```
Comment (riemass) [inline libcaf_net/src/detail/rfc6455.cpp:90]: @Neverlord  This logic could also go to framing::consume? Only thing I think it's important is to detect it early and stop reading the payload.

File: libcaf_net/src/net/web_socket/lower_layer.cpp [hunk 2]
```diff
@@ -19,7 +20,11 @@ void lower_layer::shutdown() {
 }
 
 void lower_layer::shutdown(const error& reason) {
-  shutdown(status::unexpected_condition, to_string(reason));
+  if (reason.code() == static_cast<uint8_t>(sec::protocol_error)) {
```
Comment (riemass) [inline libcaf_net/src/net/web_socket/lower_layer.cpp:23]: There are probably more edge cases I plan to add here on the go.

File: libcaf_net/test/net-test.cpp [hunk 3]
```diff
@@ -82,7 +82,6 @@ ptrdiff_t mock_stream_transport::handle_input() {
     if (consumed < 0) {
       // Negative values indicate that the application encountered an
       // unrecoverable error.
-      up->abort(make_error(caf::sec::runtime_error, "consumed < 0"));
```
Comment (riemass) [inline libcaf_net/test/net-test.cpp:85]: I wanted to check weather removing this is OK, since the framing layer will call abort on it's own. This just makes calling it twice with different error codes, so testing is harder.

File: libcaf_net/test/net/web_socket/framing.cpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+#include "caf/net/web_socket/framing.hpp"
+
+#include "net-test.hpp"
+
+using namespace caf;
+using namespace std::literals;
+
+namespace {
+
+class app_t : public net::web_socket::upper_layer {
```
Comment (riemass) [inline libcaf_net/test/net/web_socket/framing.cpp:16]: This is almost the same app_t as in client test. Only difference is in the abort call detection. I could elevate this to net-test? It seems different from the app_t in server tests, but I would really like to see only one app layer fake. Do you think it would be possible?

File: libcaf_net/test/net/web_socket/framing.cpp [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    uut = uut_layer.get();
+    transport = mock_stream_transport::make(std::move(uut_layer));
+  }
+
+  static const_byte_span make_test_data(uint64_t requested_size) {
+    static std::vector<std::byte> bytes{150, std::byte{0xFF}};
+    if (requested_size > bytes.size())
+      bytes.resize(requested_size, std::byte{0xFF});
+    return const_byte_span{bytes.data(), requested_size};
+  }
```
Comment (riemass) [inline libcaf_net/test/net/web_socket/framing.cpp:81]: I don't know if static test data is acceptable, maybe I could elevate this into net-test and use it in rfc6455 tests? Or just to implement it in that file as well?

File: libcaf_net/test/net/web_socket/framing.cpp [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+} // namespace
+
+BEGIN_FIXTURE_SCOPE(fixture)
+
+SCENARIO("the client sends a ping and receives a pong response") {
+  GIVEN("a valid WebSocket connection") {
+    std::vector<std::byte> ping_frame;
+    std::vector<std::byte> pong_frame;
+    CHECK_EQ(transport->start(nullptr), error{});
+    transport->configure_read(caf::net::receive_policy::up_to(2048u));
```
Comment (riemass) [inline libcaf_net/test/net/web_socket/framing.cpp:93]: These two lines repeat in every test. Could I put them in the fixture ctor?

File: libcaf_net/src/detail/rfc6455.cpp (hunk 1 above)
Comment (Neverlord) [reply libcaf_net/src/detail/rfc6455.cpp:90]: Intuitively it makes sense to have this here to have it centrally. But no preference either way.

File: libcaf_net/test/net/web_socket/framing.cpp (hunk 4 above)
Comment (Neverlord) [reply libcaf_net/test/net/web_socket/framing.cpp:16]: In general, a bit of duplication is OK in tests. However, if this exact app is used in multiple tests, feel free to move it to separate hpp/cpp files as a followup.

File: libcaf_net/test/net/web_socket/framing.cpp (hunk 6 above)
Comment (Neverlord) [reply libcaf_net/test/net/web_socket/framing.cpp:93]: If all of the tests will do exact this setup, sure. Makes the setup a bit less flexible, though.

File: libcaf_net/test/net/web_socket/framing.cpp [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+TEST_CASE("calling shutdown with protocol_error sets status in close header") {
+  CHECK_EQ(transport->start(nullptr), error{});
+  transport->configure_read(caf::net::receive_policy::up_to(2048u));
+  uut->shutdown(make_error(sec::protocol_error));
+  detail::rfc6455::header hdr;
+  detail::rfc6455::decode_header(transport->output_buffer(), hdr);
+  CHECK_EQ(hdr.opcode, detail::rfc6455::connection_close);
+  CHECK(hdr.payload_len >= 2);
+  auto error_code = (static_cast<uint16_t>(transport->output_buffer()[2]) << 8)
+                    + static_cast<uint16_t>(transport->output_buffer()[3]);
```
Comment (Neverlord) [inline libcaf_net/test/net/web_socket/framing.cpp:147]: 1) Since we only use this to compare values, I wouldn't bother with the uint16 type.
2) I think using `vector::at` here makes the code a bit easier to read. The bound checks are also nice to have in tests to get exceptions rather than ASAN traces.

```suggestion
  auto error_code = (transport->output_buffer().at(2) << 8)
                    + transport->output_buffer().at(3);
```

File: libcaf_net/test/net/web_socket/framing.cpp [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+  CHECK_EQ(error_code,
+           static_cast<uint16_t>(net::web_socket::status::protocol_error));
```
Comment (Neverlord) [inline libcaf_net/test/net/web_socket/framing.cpp:149]:
```suggestion
  CHECK_EQ(error_code,
           static_cast<int>(net::web_socket::status::protocol_error));
```

File: libcaf_net/test/net/web_socket/framing.cpp [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+        detail::rfc6455::header hdr;
+        detail::rfc6455::decode_header(transport->output_buffer(), hdr);
+        MESSAGE("Buffer: " << transport->output_buffer());
+        CHECK_EQ(hdr.opcode, detail::rfc6455::connection_close);
+        CHECK(hdr.payload_len >= 2);
+        auto error_code
+          = (static_cast<uint16_t>(transport->output_buffer()[2]) << 8)
+            + static_cast<uint16_t>(transport->output_buffer()[3]);
+        CHECK_EQ(error_code, static_cast<uint16_t>(
+                               net::web_socket::status::protocol_error));
```
Comment (Neverlord) [inline libcaf_net/test/net/web_socket/framing.cpp:178]: Same here: we don't need to take care of preserving uint16.

**Review by Neverlord — CHANGES_REQUESTED** [review]
Mostly nitpicks, looks good otherwise.

File: libcaf_net/test/net/web_socket/framing.cpp (hunk 9 above)
Comment (riemass) [reply libcaf_net/test/net/web_socket/framing.cpp]: Right, we can use `int`. `static_cast` is still needed since we need to explicitly create `int` it from `std::byte`.

Comment (Neverlord) [reply libcaf_net/test/net/web_socket/framing.cpp]: Ah, if this is a `std::byte`: the idiomatic conversion would be [`to_integer`](https://en.cppreference.com/w/cpp/types/byte).

File: libcaf_net/caf/detail/rfc6455.hpp [hunk 10]
```diff
@@ -57,6 +57,10 @@ struct CAF_NET_EXPORT rfc6455 {
                              uint8_t flags = fin_flag);
 
   static ptrdiff_t decode_header(const_byte_span data, header& hdr);
+
+  static bool is_control_frame(uint8_t opcode) {
```
[the remark this answers is not in the document]
Comment (riemass) [reply libcaf_net/caf/detail/rfc6455.hpp]: I missed that.

File: libcaf_net/caf/detail/rfc6455.hpp [hunk 11]
```diff
@@ -57,6 +57,10 @@ struct CAF_NET_EXPORT rfc6455 {
                              uint8_t flags = fin_flag);
 
   static ptrdiff_t decode_header(const_byte_span data, header& hdr);
+
+  static constexpr bool is_control_frame(uint8_t opcode) {
```
Comment (Neverlord) [inline libcaf_net/caf/detail/rfc6455.hpp:61]: The `noexcept` got lost from my last suggestion:

```suggestion
  static constexpr bool is_control_frame(uint8_t opcode) noexcept {
```

**Review by Neverlord — CHANGES_REQUESTED** [review]

**Review by Neverlord — APPROVED** [review]

File: libcaf_net/caf/detail/rfc6455.hpp (hunk 11 above)
Comment (riemass) [reply libcaf_net/caf/detail/rfc6455.hpp]: Sorry about that. Should be fine now.
